0000-classical-stores.md

```diff
@@ -0,0 +1,493 @@
+
+This is a crude approximation at proper SSA tracking of variables, but this should be sufficient, if a little over-cautious; certain assignments in deeply nested conditions could introduce ordering relations that could be somewhat simplified.
+In general, though, there's only so much Qiskit should attempt at this point.
+
+
+### The `Store` instruction
+
+`QuantumCircuit.store` is a shorthand for emitting a `Store` subclass of `Instruction`.
+For further discussion on why this is a subclass of `Instruction` and not a more fundamental modification to `QuantumCircuit.data` or `CircuitInstruction`, see the "alternatives" section below.
+
+The two parameters of `Store` are the location and the value.
+These are stored in `Store.params`, like a parametrised gate object, so that they can be moved to a more suitable location in a similar manner in the future.
+It is not expected that any code that currently inspects `Instruction.params` should be able to automatically work with `Store`; these objects require special handling, but this is generally true for the `.params` field which is (unfortunately) quite unconstrained in its typing.
+
+
+## Alternative Approaches
+
+### Large-scale alternatives
+
+#### Expand `condition` or `Expr` fields to support `ParameterExpression` and use `assign_parameters`
+
+As discussed in [RFC 10](0010-simple-classical-representations.md), reuse of `ParameterExpression` is not suitable for typed inputs.
+It also fundamentally represents a symbolic calculation on real numbers, rather than a circuit run-time evaluation of an object, so it would be close to impossible to align the two computational models.
+
+#### Give `Var` nodes backing `Clbit`s
+
+I seriously considered this to the degree of holding up publishing this RFC to pivot to this design.
+It has the benefit of:
+- making `ClassicalRegister` a little less special-cased; it's a variable with contained bits.
+- it would be possible to represent a `measure` directly into a bit of a new-style variable
+
+However, in the end, I have many reasons that I think this would not the correct decision:
+
+- When 
+- There's 5 pointers of storage space required per Python-space `Clbit` that's just wrapped by a `Var`, not to mention the additional costs of tracking the bit within `QuantumCircuit`.
+  For a `double`, that's over 2kB of extra tracking, even if we do nothing with it.
+- It implies that we need to track each of these `Clbit` wires separately in the DAG, even though the data-flow is bundled - again, lots of unnecessary memory use.
+- There is a natural split in hardware between a `Clbit`, which is a target for a `measure` instruction and is likely written to in controllers very close to the QPU hardware, and classical variables that can be in outer control systems.
+  For example, gathering the result of a `measure` into the particular bit of a `uint[8]` might involve more expensive transfer and synchronisation operations than adding together two `uint[8]` objects.
```

Comment:
This distinction makes me a bit uncomfortable, particularly because when you serialize a `Var` of type `bit` or a `Clbit` to OpenQASM they should end up indistinguishable.